File: frontend/apps/android/app/src/test/java/app/pantopus/android/ui/screens/gigs/tasks_map/TasksMapViewModelTests.kt
@file:Suppress("PackageNaming", "MagicNumber")

package app.pantopus.android.ui.screens.gigs.tasks_map

import androidx.lifecycle.SavedStateHandle
import app.pantopus.android.data.api.models.gigs.GigDto
import app.pantopus.android.data.api.models.gigs.GigsInBoundsResponse
import app.pantopus.android.data.api.net.NetworkError
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.gigs.GigsRepository
import app.pantopus.android.ui.screens.gigs.GigsCategory
import app.pantopus.android.ui.screens.gigs.GigsSort
import app.pantopus.android.ui.screens.shared.map_list_hybrid.MapPinState
import io.mockk.coEvery
import io.mockk.mockk
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.UnconfinedTestDispatcher
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.test.setMain
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

/**
 * A11.1 Tasks map view-model — live `GET /api/gigs/in-bounds` fetch +
 * projection, plus the client-side category filter, pin↔card selection,
 * and sort. Mirrors iOS `TasksMapViewModelTests`.
 */
@OptIn(ExperimentalCoroutinesApi::class)
class TasksMapViewModelTests {
    private val repo: GigsRepository = mockk()

    @Before fun setUp() {
        Dispatchers.setMain(UnconfinedTestDispatcher())
    }

    @After fun tearDown() {
        Dispatchers.resetMain()
    }

    // Anchor is TasksMapSampleData.anchor (40.7484, -73.9857); coords are
    // placed so handyman-1 is the closest task.
    private fun gig(
        id: String,
        category: String,
        price: Double,
        lat: Double,
        lon: Double,
        bidCount: Int,
        status: String = "open",
        payType: String? = null,
    ): GigDto =
        GigDto(
            id = id,
            title = "Task $id",
            price = price,
            category = category,
            status = status,
            payType = payType,
            bidCount = bidCount,
            latitude = lat,
            longitude = lon,
        )

    private val sampleGigs =
        listOf(
            gig("handyman-1", "handyman", 60.0, 40.749, -73.988, bidCount = 4),
            gig("cleaning-1", "cleaning", 180.0, 40.745, -73.982, bidCount = 1),
            gig("cleaning-2", "cleaning", 90.0, 40.752, -73.990, bidCount = 2),
            gig("petcare-1", "petcare", 22.0, 40.747, -73.984, bidCount = 0, payType = "per_walk"),
            gig("petcare-2", "petcare", 25.0, 40.750, -73.987, bidCount = 3, payType = "per_walk"),
            gig("moving-1", "moving", 80.0, 40.744, -73.983, bidCount = 5, status = "assigned"),
            gig("tutoring-1", "tutoring", 40.0, 40.753, -73.991, bidCount = 6, status = "in_progress"),
        )

    private fun vm(
        category: String? = null,
        gigs: List<GigDto> = sampleGigs,
        result: NetworkResult<GigsInBoundsResponse> = NetworkResult.Success(GigsInBoundsResponse(gigs)),
    ): TasksMapViewModel {
        coEvery { repo.inBounds(any(), any(), any(), any(), any()) } returns result
        return TasksMapViewModel(
            repo,
            SavedStateHandle(if (category != null) mapOf("category" to category) else emptyMap()),
        )
    }

    private fun items(vm: TasksMapViewModel): List<TaskMapItem>? = (vm.state.value as? TasksMapUiState.Populated)?.items

    @Test
    fun load_fetches_in_bounds_and_projects_with_closest_selection() =
        runTest {
            val vm = vm()
            vm.load()
            assertEquals(sampleGigs.size, items(vm)?.size)
            // Closest-first by default → handyman-1 leads + pulses.
            assertEquals("handyman-1", vm.selectedId.value)
            assertEquals(items(vm)?.first()?.id, vm.selectedId.value)
            assertEquals("$60", items(vm)?.first()?.price)
            assertEquals("$22/walk", items(vm)?.firstOrNull { it.id == "petcare-1" }?.price)
        }

    @Test
    fun load_drops_gigs_without_coordinates() =
        runTest {
            val noCoords = GigDto(id = "x", title = "No coords", category = "cleaning", status = "open")
            val vm = vm(gigs = listOf(sampleGigs.first(), noCoords))
            vm.load()
            assertEquals(listOf("handyman-1"), items(vm)?.map { it.id })
        }

    @Test
    fun load_empty_result_produces_empty() =
        runTest {
            val vm = vm(gigs = emptyList())
            vm.load()
            assertTrue(vm.state.value is TasksMapUiState.Empty)
            assertNull(vm.selectedId.value)
        }

    @Test
    fun load_server_error_produces_error() =
        runTest {
            val vm = vm(result = NetworkResult.Failure(NetworkError.Server(500, null)))
            vm.load()
            assertTrue(vm.state.value is TasksMapUiState.Error)
        }

    @Test
    fun category_filter_narrows_visible_items() =
        runTest {
            val vm = vm()
            vm.load()
            vm.selectCategory(GigsCategory.Cleaning)
            val visible = items(vm)
            assertEquals(2, visible?.size)
            assertTrue(visible?.all { it.category == GigsCategory.Cleaning } == true)
            assertEquals(visible?.first()?.id, vm.selectedId.value)
        }

    @Test
    fun category_with_no_matches_produces_empty() =
        runTest {
            val vm = vm()
            vm.load()
            vm.selectCategory(GigsCategory.Tech) // no tech task fetched
            assertTrue(vm.state.value is TasksMapUiState.Empty)
            assertNull(vm.selectedId.value)
        }

    @Test
    fun widen_from_empty_restores_populated() =
        runTest {
            val vm = vm()
            vm.load()
            vm.selectCategory(GigsCategory.Tech)
            assertTrue(vm.state.value is TasksMapUiState.Empty)
            vm.selectCategory(GigsCategory.All) // "Widen search"
            assertEquals(sampleGigs.size, items(vm)?.size)
        }

    @Test
    fun select_updates_selected_id() =
        runTest {
            val vm = vm()
            vm.load()
            vm.select("cleaning-1")
            assertEquals("cleaning-1", vm.selectedId.value)
        }

    @Test
    fun sort_fewest_bids_orders_ascending() =
        runTest {
            val vm = vm()
            vm.load()
            vm.selectSort(GigsSort.FewestBids)
            val bids = items(vm)?.map { it.bidCount } ?: emptyList()
            assertEquals(bids.sorted(), bids)
        }

    @Test
    fun sort_highest_pay_leads_with_priciest() =
        runTest {
            val vm = vm()
            vm.load()
            vm.selectSort(GigsSort.HighestPay)
            assertEquals("cleaning-1", items(vm)?.first()?.id) // $180
        }

    @Test
    fun initial_category_from_saved_state_applied_on_load() =
        runTest {
            val vm = vm(category = "petcare")
            vm.load()
            val visible = items(vm)
            assertEquals(2, visible?.size)
            assertTrue(visible?.all { it.category == GigsCategory.PetCare } == true)
            assertEquals(GigsCategory.PetCare, vm.activeCategory.value)
        }

    @Test
    fun pins_carry_pending_state_for_non_open_gigs() =
        runTest {
            val vm = vm()
            vm.load()
            val pins = items(vm)?.map { it.toPin() } ?: emptyList()
            assertEquals(sampleGigs.size, pins.size)
            // moving-1 (assigned) + tutoring-1 (in_progress) → pending.
            assertEquals(2, pins.count { it.state == MapPinState.Pending })
        }
}

File: frontend/apps/ios/PantopusTests/Features/Gigs/TasksMapViewModelTests.swift
//
//  TasksMapViewModelTests.swift
//  PantopusTests
//
//  A11.1 Tasks map view-model — the live `GET /api/gigs/in-bounds` fetch
//  + projection, plus the client-side category filter, pin↔card
//  selection, and sort that run on the fetched (or seeded) set. The
//  filter/sort/selection tests drive an explicit `seed` (offline mode);
//  the fetch tests drive a stubbed `APIClient`.
//

import XCTest
@testable import Pantopus

@MainActor
final class TasksMapViewModelTests: XCTestCase {
    override func setUp() {
        super.setUp()
        SequencedURLProtocol.reset()
    }

    private func makeAPI() -> APIClient {
        APIClient(
            environment: .current,
            session: SequencedURLProtocol.makeSession(),
            retryPolicy: .none
        )
    }

    private func seeded(_ initialCategory: GigsCategory = .all) -> TasksMapViewModel {
        TasksMapViewModel(initialCategory: initialCategory, seed: TasksMapSampleData.items)
    }

    private func items(_ state: TasksMapState) -> [TaskMapItem]? {
        if case let .populated(items) = state { return items }
        return nil
    }

    private func isEmpty(_ state: TasksMapState) -> Bool {
        if case .empty = state { return true }
        return false
    }

    private func errorMessage(_ state: TasksMapState) -> String? {
        if case let .error(message) = state { return message }
        return nil
    }

    // MARK: - Live in-bounds fetch

    func testLoadFetchesInBoundsAndProjectsPins() async {
        // Two gigs near the default anchor (40.7484, -73.9857); g1 sits
        // closer so closest-sort (the default) leads with it.
        SequencedURLProtocol.sequence = [
            .status(200, body: """
            {"gigs":[
              {"id":"g1","title":"Fix a leaky sink","category":"handyman","price":60,
               "pay_type":"fixed","bid_count":4,"status":"open",
               "latitude":40.749,"longitude":-73.988},
              {"id":"g2","title":"Walk my dog","category":"pet","price":22,
               "pay_type":"per_walk","bid_count":1,"status":"open",
               "latitude":40.740,"longitude":-73.978}
            ]}
            """)
        ]
        let vm = TasksMapViewModel(api: makeAPI())
        await vm.load()
        let visible = items(vm.state)
        XCTAssertEqual(visible?.count, 2)
        XCTAssertEqual(visible?.first?.id, "g1", "Closest task leads + pulses")
        XCTAssertEqual(vm.selectedId, "g1")
        XCTAssertEqual(visible?.first?.category, .handyman)
        XCTAssertEqual(visible?.first?.price, "$60")
        XCTAssertEqual(visible?.first(where: { $0.id == "g2" })?.price, "$22/walk")
        XCTAssertEqual(visible?.first(where: { $0.id == "g2" })?.category, .petcare)
        // Distance was computed client-side, not "—".
        XCTAssertNotEqual(visible?.first?.distanceLabel, "—")
    }

    func testLoadDropsGigsWithoutCoordinates() async {
        SequencedURLProtocol.sequence = [
            .status(200, body: """
            {"gigs":[
              {"id":"g1","title":"Has coords","category":"handyman","price":50,"status":"open",
               "latitude":40.749,"longitude":-73.988},
              {"id":"g2","title":"No coords","category":"cleaning","price":50,"status":"open"}
            ]}
            """)
        ]
        let vm = TasksMapViewModel(api: makeAPI())
        await vm.load()
        XCTAssertEqual(items(vm.state)?.map(\.id), ["g1"])
    }

    func testLoadEmptyResultProducesEmpty() async {
        SequencedURLProtocol.sequence = [.status(200, body: "{\"gigs\":[]}")]
        let vm = TasksMapViewModel(api: makeAPI())
        await vm.load()
        XCTAssertTrue(isEmpty(vm.state))
        XCTAssertNil(vm.selectedId)
    }

    func testLoadServerErrorProducesError() async {
        SequencedURLProtocol.sequence = [.status(500, body: "{\"error\":\"boom\"}")]
        let vm = TasksMapViewModel(api: makeAPI())
        await vm.load()
        XCTAssertNotNil(errorMessage(vm.state))
    }

    // MARK: - Offline seed: states + selection

    func testLoadProducesPopulatedWithDefaultSelection() async {
        let vm = seeded()
        await vm.load()
        XCTAssertEqual(items(vm.state)?.count, TasksMapSampleData.items.count)
        // Closest-first by default → the 0.2 mi handyman task leads + pulses.
        XCTAssertEqual(vm.selectedId, items(vm.state)?.first?.id)
        XCTAssertEqual(vm.selectedId, "handyman-1")
    }

    func testLoadEmptySeedProducesEmpty() async {
        let vm = TasksMapViewModel(seed: [])
        await vm.load()
        XCTAssertTrue(isEmpty(vm.state))
        XCTAssertNil(vm.selectedId)
    }

    func testLoadFailureProducesError() async {
        let vm = TasksMapViewModel(seed: TasksMapSampleData.items, failWith: "Couldn't load tasks")
        await vm.load()
        XCTAssertEqual(errorMessage(vm.state), "Couldn't load tasks")
    }

    func testCategoryFilterNarrowsVisibleItems() async {
        let vm = seeded()
        await vm.load()
        vm.selectCategory(.cleaning)
        let visible = items(vm.state)
        XCTAssertEqual(visible?.count, 2)
        XCTAssertEqual(visible?.allSatisfy { $0.category == .cleaning }, true)
        // Selection follows the filter to a visible task.
        XCTAssertEqual(vm.selectedId, visible?.first?.id)
    }

    func testCategoryWithNoMatchesProducesEmpty() async {
        let vm = seeded()
        await vm.load()
        vm.selectCategory(.tech) // no tech task in the seed
        XCTAssertTrue(isEmpty(vm.state))
        XCTAssertNil(vm.selectedId)
    }

    func testWidenFromEmptyRestoresPopulated() async {
        let vm = seeded()
        await vm.load()
        vm.selectCategory(.tech)
        XCTAssertTrue(isEmpty(vm.state))
        vm.selectCategory(.all) // "Widen search"
        XCTAssertEqual(items(vm.state)?.count, TasksMapSampleData.items.count)
    }

    func testSelectUpdatesSelectedId() async {
        let vm = seeded()
        await vm.load()
        vm.select("cleaning-1")
        XCTAssertEqual(vm.selectedId, "cleaning-1")
    }

    func testSortFewestBidsOrdersAscending() async {
        let vm = seeded()
        await vm.load()
        vm.selectSort(.fewestBids)
        let bids = items(vm.state)?.map(\.bidCount) ?? []
        XCTAssertEqual(bids, bids.sorted())
    }

    func testSortHighestPayLeadsWithPriciestTask() async {
        let vm = seeded()
        await vm.load()
        vm.selectSort(.highestPay)
        XCTAssertEqual(items(vm.state)?.first?.id, "cleaning-1") // $180
    }

    func testInitialCategoryAppliedOnLoad() async {
        let vm = seeded(.petcare)
        await vm.load()
        let visible = items(vm.state)
        XCTAssertEqual(visible?.count, 2)
        XCTAssertEqual(visible?.allSatisfy { $0.category == .petcare }, true)
        XCTAssertEqual(vm.activeCategory, .petcare)
    }

    func testPinProjectionCarriesCategoryColorAndState() async {
        let vm = seeded()
        await vm.load()
        let pins = items(vm.state)?.map(\.pin) ?? []
        XCTAssertEqual(pins.count, TasksMapSampleData.items.count)
        // The moving + tutoring seeds are pending; the rest confirmed.
        let pending = pins.filter { $0.state == .pending }.count
        XCTAssertEqual(pending, 2)
    }
}
